src/libisyntax.h

```diff
@@ -1,3 +1,68 @@
 #pragma once
 
-void libisyntax_init();
+#include <stdint.h>
+
+// API conventions:
+// - return type is one of:
+// -- appropriate data type for simple getters. For complex getters doing work, isyntax_error_t + out variable.
+// -- void if the function doesn't fail at runtime (e.g. destructors).
+// -- isyntax_error_t otherwise.
+// - Output arguments are last, name prefixed with 'out_'.
+// - 'Object operated on' argument is first.
+// - Function names are prefixed with 'libisyntax_', as C doesn't have namespaces. Naming convention
+//   is: libisyntax_<object>_<action>(..), where <object> is omitted for the object representing the isyntax file.
+// - Functions don't check for null/out of bounds, for efficiency. (they may assert, but that is implementation detail).
+// - Booleans are represented as int32_t and prefxied with 'is_' or 'has_'.
+// - Const applied to pointers is used as a signal that the object will not be modified.
+// - Prefer int even for unsigned types, see java rationale.
+
+
+typedef int32_t isyntax_error_t;
+
+#define LIBISYNTAX_OK 0
+// Generic error that the user should not expect to recover from.
+#define LIBISYNTAX_FATAL 1
+// One of the arguments passed to a function is invalid.
+#define LIBISYNTAX_INVALID_ARGUMENT 2
+
+typedef struct isyntax_t isyntax_t;
+typedef struct isyntax_image_t isyntax_image_t;
+typedef struct isyntax_level_t isyntax_level_t;
+typedef struct isyntax_cache_t isyntax_cache_t;
+
+//== Common API ==
+isyntax_error_t libisyntax_init();
```

Comment:
The only safe way to call this from another library is from that library's shared-library constructor, because there's no way to synchronize init calls between multiple libraries that both use libisyntax.  Would it make sense to remove this from the API and either a) call it from libisyntax's shared library constructor or b) call it from library entry points (`libisyntax_open` and `libisyntax_cache_create`) using proper locking?

Reply:
Just to check, I think we could do a global mutex within libisyntax_init() and declare it thread-safe, right?
In OpenSlide integration, I currently call it from a static-mutex-locked block, but we can move that static-mutex into libisyntax_init(). Or would that not work for some case?

More generally (from my latest code):
// TODO(avirodov): is libisyntax_init thread-safe? Maybe need a general api-level statement about thread-safety.

Reply:
If `libisyntax_init()` uses its own internal locking and is safe to call more than once, that should be fine.  But in that case, is there any reason to expose it in the API, or should it be called implicitly as in (b) above?